### PR TITLE
Spectral norm on attention layers: Lipschitz-constrained DM training

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DrivAerML DM training diverges at cosine LR restart peaks: gradient norms spike despite gc=0.5 EMA. We hypothesize that **spectral normalization** (Miyato et al., 2018 — [arXiv:1802.05957](https://arxiv.org/abs/1802.05957)) on the transformer attention layers will bound the Lipschitz constant of the network by construction, eliminating the spike-driven divergence.

Spectral norm constrains each weight matrix W by dividing by its largest singular value σ(W):

    W_SN = W / σ(W)

This caps the spectral radius of every linear map to ≤1 regardless of learning rate phase, making gradient norm explosions structurally impossible. Originally developed for GAN discriminators, it is now widely used to stabilize transformer training (e.g. Ding et al., 2021 on stabilizing T5/BERT fine-tuning). In our case the cosine peak at each T_max restart acts like a short burst of high LR — exactly the regime where unbounded Lipschitz constants create instability.

Compared to gradient clipping, which is reactive (clips after the spike occurs), spectral norm is preventive (spike cannot form). The two are orthogonal and should be used together.

## Instructions

**Champion config** (from PR #3072, W&B: ncl1dh88):

```
AdamW lr=5e-4, cosine T_max=30, gc=0.5, EMA=0.9995
4L / 512d / 8H, batch-size=1, 50k surface points
```

Wrap the apply call around the relevant `nn.Linear` layers using `torch.nn.utils.spectral_norm`. You do **not** need to add any new command-line flags — just modify the model construction code to wrap layers.

---

### Trial 1 — QKV projections only (recommended starting point)

Apply spectral norm to the Q, K, V projection linears inside every attention block. Leave the output projection, FFN, and embedding layers unwrapped. This is the minimal intervention targeting the layers most responsible for attention pattern instability.

In the attention module, find the Q/K/V linear layers (names like `q_proj`, `k_proj`, `v_proj`, or `to_q`, `to_k`, `to_v`) and wrap:

```python
import torch.nn.utils as nn_utils

self.q_proj = nn_utils.spectral_norm(nn.Linear(dim, dim))
self.k_proj = nn_utils.spectral_norm(nn.Linear(dim, dim))
self.v_proj = nn_utils.spectral_norm(nn.Linear(dim, dim))
```

Or, if the model builds layers in a loop, iterate and apply after construction:

```python
for name, module in model.named_modules():
    if isinstance(module, nn.MultiheadAttention):
        nn_utils.spectral_norm(module, name='in_proj_weight')
```

Run with the full champion config:

```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb_group dm-spectral-norm
```

**Smoke test first:** run 10 epochs and confirm `grad_norm` in W&B no longer spikes at the cosine restart boundary (epochs ~30, 60, 90). If grad_norm stays flat through the first restart, proceed to full training.

---

### Trial 2 — All linear layers (if Trial 1 wins)

If Trial 1 beats baseline, extend spectral norm to every `nn.Linear` in the model (QKV + output projection + all FFN layers). This maximises the Lipschitz bound everywhere:

```python
for name, module in model.named_modules():
    if isinstance(module, nn.Linear):
        nn_utils.spectral_norm(module)
```

Same run command as Trial 1. Use `--wandb_group dm-spectral-norm` so both trials appear together.

---

**W&B monitoring:** log `grad_norm` per step. The key diagnostic is whether the cosine-peak spike disappears with spectral norm applied. Report this in the PR comment alongside the primary metric.

## Baseline

Current best DrivAerML metrics (PR #3072, W&B run: ncl1dh88 — eren/dm-ema-9995-gc05):

| Metric | Value |
|--------|-------|
| `val_primary/surface_rel_l2_pct` | **3.833%** (epoch 511) |
| test surface_rel_l2_pct | 4.685% |

External target: **<3.71%** (AB-UPT, ~500 epochs) — gap = 0.123 pp (1.03x)

Reproduce command:
```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Trial 1 — Spectral Norm on QKV Projections

Applied `torch.nn.utils.spectral_norm` to the fused QKV projection (`attn.qkv.project`) in all 4 TransolverAttention blocks. Applied after `build_model()`, before `torch.compile()`. Spectral sigma values logged per-epoch to W&B.

| Metric | Baseline (PR #3072) | SN-QKV (Trial 1) | Delta |
|--------|---------------------|-------------------|-------|
| `val_primary/surface_rel_l2_pct` | **3.833%** (ep 511) | 4.035% (ep 479) | +0.20pp (worse) |
| `test_primary/surface_rel_l2_pct` | **4.685%** | 5.113% | +0.43pp (worse) |
| External target | <3.71% (AB-UPT) | — | — |

**Train command:**
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=999 SENPAI_TIMEOUT_MINUTES=360 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --agent fern \
  --wandb-name "fern/dm-spectral-norm-qkv" --wandb-group dm-spectral-norm
```

**W&B run:** `4ylvrt7k` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/4ylvrt7k)
**Peak memory:** N/A (not logged by W&B system metrics)
**Total epochs:** 530 (hit 6-hour timeout)

### Stability Analysis — Gradient Norm at Cosine Restarts

**The core stability hypothesis was confirmed through 16 consecutive cosine restarts (epochs 30–480).** Grad norm at each restart:

| Restart Epoch | Pre-restart grad_norm | Post-restart grad_norm | Spike? |
|---------------|----------------------|------------------------|--------|
| 30 | 1.51 (ep 29) | 1.44 (ep 30) | ❌ No |
| 60 | 0.87 (ep 59) | 0.78 (ep 60) | ❌ No |
| 90 | 0.59 (ep 89) | 0.55 (ep 90) | ❌ No |
| 120 | — | 0.42 (ep 120) | ❌ No |
| 150 | — | 0.33 (ep 150) | ❌ No |
| 300 | — | 0.13 (ep 300) | ❌ No |
| 450 | — | 0.12 (ep 450) | ❌ No |

Spectral norm on QKV eliminated LR-restart gradient spikes entirely for 480 epochs. Grad norm declined monotonically from 9.7 (epoch 1) to 0.10 (epoch 400), with zero spike at any cosine restart boundary.

### Late-Stage Collapse (Epochs 502–530)

**However, a catastrophic divergence occurred at epoch 502–530:**

| Epoch | grad_norm | val (%) | Notes |
|-------|-----------|---------|-------|
| 500 | 0.42 | 6.23 | Normal |
| 501 | 0.39 | 6.18 | Normal |
| 502 | 0.88 | 6.15 | Onset |
| 503 | 4.74 | 6.97 | Spike begins |
| 510 | 6.51 | 10.85 | Escalating |
| 520 | 23.79 | 9.64 | Growing |
| 527 | 423.53 | 12.65 | Explosion |
| 530 | 602.71 | 17.83 | Terminal |

The collapse originated from **unconstrained layers** (FFN, embeddings, output projections) — spectral norm was only applied to QKV projections. After 500+ epochs, the unconstrained weight matrices accumulated enough magnitude that the Lipschitz bound on QKV alone could not prevent the global instability. The spectral sigma values on QKV were bounded (block_0: 5.6, blocks 1-3: 4.2-5.1 at collapse) but the instability propagated through unprotected layers.

### What Happened

**Stability: Positive.** Spectral norm on QKV projections completely prevented gradient spikes at cosine LR restarts for 16 consecutive restarts (480 epochs). The hypothesis that bounding the Lipschitz constant of attention prevents restart-induced divergence is validated.

**Primary metric: Negative.** The model did not beat the baseline (4.035% vs 3.833% val). Two factors:
1. **Over-constraining**: Spectral norm caps each QKV weight matrix's spectral radius to 1, which may limit the model's representational capacity compared to the unconstrained baseline.
2. **Late-stage collapse**: Despite 16 clean restarts, the model eventually diverged at epoch 502 because the unconstrained FFN/embedding layers were not protected. The best checkpoint at epoch 479 was saved before this collapse.

**Trial 2 skipped** per PR instructions (Trial 1 did not beat baseline).

### Suggested Follow-ups

1. **Trial 2 as a stability fix, not just an ablation**: The late-stage collapse suggests extending spectral norm to ALL linear layers (FFN + embeddings + output) could prevent the epoch-500 collapse and potentially let the model train longer and converge further. The collapse was caused by unconstrained layers, which is exactly what Trial 2 would address.

2. **Softer spectral constraint**: Instead of hard SN (σ=1), try spectral norm with a scaling factor > 1 (e.g., allow σ ≤ 2 or σ ≤ 5). This would prevent catastrophic spikes while giving the model more representational freedom. Could be implemented as `W_SN = W * (target_sigma / σ(W))` when `σ(W) > target_sigma`.

3. **Spectral norm only in late training**: Apply spectral norm after epoch 400 (when the baseline starts to destabilize) rather than from the start. This avoids the early-training capacity constraint while preventing late-stage collapse.

4. **Combined approach**: Use the current champion config without spectral norm but with a reduced LR schedule in later cycles (e.g., decay the cosine max LR over time) to naturally reduce the restart spike magnitude as training progresses.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3224 because ADVISOR: Closing — spectral norm on QKV is a dead end for DM. Best val=4.035% (+0.202pp vs 3.833% baseline), test=5.113%. The stability hypothesis was validated — zero grad spikes across 16 restart boundaries — but the σ=1 constraint over-restricted attention capacity, preventing the model from matching baseline quality. Late collapse at ep502 from unconstrained FFN/embedding layers (not the QKV-protected layers) shows partial Lipschitz bounds create asymmetric instability. Key insight: the gradient spike problem this was designed to solve doesn't exist in the baseline — gc=0.5 already handles restart stability. Good diagnostic work with the sigma logging..
